### PR TITLE
XSD validation for Lucene datastore

### PR DESCRIPTION
**Pull request by mgov88**
I implemented an XSD validation for the Lucene Datastore, I would like to know if I did it correctly, so I can go on and repeat this implementation for the other datastores.

I created this PR as a draft, when I finish the rest of datastores I will send a complete PR.

Issue: [GORA-455 Create XSD's for all datastore xml mapping files](https://issues.apache.org/jira/projects/GORA/issues/GORA-455)

### CODE REVIEW

File: gora-lucene/src/test/conf/gora-lucene.xsd [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
```
Comment (jhnmora000) [inline gora-lucene/src/test/conf/gora-lucene.xsd:2]: please add the Apache 2.0 License notice for this file

File: gora-lucene/src/test/conf/gora-lucene.xsd [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    <xs:element name="gora-otd" type="lucene-mapping"/>
```
Comment (jhnmora000) [inline gora-lucene/src/test/conf/gora-lucene.xsd:4]: format this file with 2-space indentation

File: gora-lucene/src/test/conf/gora-lucene.xsd [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (jhnmora000) [inline gora-lucene/src/test/conf/gora-lucene.xsd:1]: IMHO, a schema file should not be placed in the tests of the datastore

**Review by jhnmora000 — CHANGES_REQUESTED** [review]
please address these comments before you start working on other datastores.

File: gora-lucene/src/test/conf/gora-lucene.xsd (hunk 1 above)
Comment (mgov88) [reply gora-lucene/src/test/conf/gora-lucene.xsd:2]: Thanks for your feedback, I will add it

File: gora-lucene/src/test/conf/gora-lucene.xsd (hunk 2 above)
Comment (mgov88) [reply gora-lucene/src/test/conf/gora-lucene.xsd:4]: Sorry , I did realised that I used the wrong indentation

File: gora-lucene/src/test/conf/gora-lucene.xsd (hunk 3 above)
Comment (mgov88) [reply gora-lucene/src/test/conf/gora-lucene.xsd:1]: I am going to move the schema file to a 'resources' folder in the main code of the module.

**Comment by alfonsonishikawa** [comment]
Would be desirable to add the XSD reference to the Lucene mapping XMLs (for sure there is some XML mapping in tests).

File: gora-lucene/src/test/conf/gora-lucene-mapping.xml [hunk 4]
```diff
@@ -16,7 +16,8 @@
    limitations under the License.
 -->
 
-<gora-otd>
+<gora-otd xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
+          xsi:noNamespaceSchemaLocation="gora-lucene.xsd">
```
Comment (alfonsonishikawa) [inline gora-lucene/src/test/conf/gora-lucene-mapping.xml:20]: What does "noNamespaceSchemaLocation" means? Is that the default namespace? Do <class> and <primarykey> use that by default?
I suggest to create a Unit Test to check that an ill-formed mapping does not pass the validation.

Comment (mgov88) [reply gora-lucene/src/test/conf/gora-lucene-mapping.xml:20]: Hi, it found examples on the internet that used that attributes to link to local XSD files, I guess this noNamespaceSchemaLocation is used if there is not a Namespaces defined in the XML.

I will try to implement a test

Comment (mgov88) [reply gora-lucene/src/test/conf/gora-lucene-mapping.xml:20]: I developed a test. When a bad XML file is parsed this exception triggers:

`org.apache.gora.util.GoraException: java.io.IOException: Unable to read gora-lucene-mapping-bad.xml
	at org.apache.gora.lucene.store.TestLuceneStore.testXSDValidation(TestLuceneStore.java:178)
Caused by: java.io.IOException: Unable to read gora-lucene-mapping-bad.xml
	at org.apache.gora.lucene.store.TestLuceneStore.testXSDValidation(TestLuceneStore.java:178)
Caused by: org.xml.sax.SAXParseException: cvc-complex-type.2.4.a: Invalid content was found starting with element 'Field'. One of '{primarykey}' is expected.
	at org.apache.gora.lucene.store.TestLuceneStore.testXSDValidation(TestLuceneStore.java:178)`

**Comment by alfonsonishikawa** [comment]
Good to see it works! :)
The last thing I would ask for is to add an opt-out (or opt-in) option to gora.properties to enable/disable verification.

**Comment by alfonsonishikawa** [comment]
Merged to master. Thank you, @mgov88 !

**Pull request merged**